Metrics: Expose underlying queue status

SimonRichardson opened the pull request:
The following exposes the underlying status of a subscriber queue. The
queue per subscriber can be susceptible to back pressure from the slow
callbacks. We currently don't have any visibility of how this manifests
as we don't have any metrics to observe. The following commit exposes
this in convenient methods that each hub can implement.

Implementations of the metrics interface should be careful around data
races, as the metrics type walks across many goroutines.

### Benchmarks

#### Pre changes

```
$ go test -v ./... -check.b
=== RUN   TestPackage
PASS: benchmarks_test.go:19: BenchmarkSuite.BenchmarkStructuredNoConversions      500000              4227 ns/op
PASS: benchmarks_test.go:43: BenchmarkSuite.BenchmarkStructuredSerialize          100000             17552 ns/op
OK: 2 passed
--- PASS: TestPackage (4.13s)
PASS
ok      github.com/juju/pubsub/v2       4.142s

```


#### Post changes

```
$ go test -v ./... -check.b
=== RUN   TestPackage
PASS: benchmarks_test.go:19: BenchmarkSuite.BenchmarkStructuredNoConversions      500000              4798 ns/op
PASS: benchmarks_test.go:43: BenchmarkSuite.BenchmarkStructuredSerialize          100000             18281 ns/op
OK: 2 passed
--- PASS: TestPackage (4.51s)
PASS
ok      github.com/juju/pubsub/v2       4.524s
```

#### Post changes with function naming

```
go test -v ./... -check.b
=== RUN   TestPackage
PASS: benchmarks_test.go:19: BenchmarkSuite.BenchmarkStructuredNoConversions      500000              4641 ns/op
PASS: benchmarks_test.go:43: BenchmarkSuite.BenchmarkStructuredSerialize          100000             18359 ns/op
OK: 2 passed
--- PASS: TestPackage (4.44s)
PASS
ok      github.com/juju/pubsub/v2       4.453s
````

SimonRichardson committed:
Metrics: Expose underlying queue status

The following exposes the underlying status of a subscriber queue. The
queue per subscriber can be susceptible to back pressure from the slow
callbacks. We currently don't have any visibility of how this manifests
as we don't have any metrics to observe. The following commit exposes
this in convenient methods that each hub can implement.

Implementations of the metrics interface should be careful around data
races, as the metrics type walks across many goroutines.

jameinel commented on metrics.go:
```diff
@@ -0,0 +1,35 @@
+	// via the callback, or how long it took.
+	Dequeued(index int)
+
+	// Consumed metric identifies the number of consumed messages the subscriber
```
You say the 'number of consumed messages' but isn't this indicating a single consumed message?

SimonRichardson replied:
Fixed

SimonRichardson committed:
Metrics: Add subscribe and unsubscribe metrics

To see how many subscriptions a hub has per hub, we need to expose the
following metric.

jameinel commented on subscriber.go:
```diff
@@ -65,40 +78,64 @@ func (s *subscriber) loop() {
+			// information to workout how much backpressure is being exhorted
+			// on the system at large.
+			s.metrics.Consumed(s.id, s.clock.Now().Sub(message.now))
 		}
```
if you did get nil, should we be reporting some sort of 'juju should be debugging this' rather than silently ignoring it? The earlier message seems very strong for dropping the condition on the floor.

SimonRichardson replied:
Agreed, I've now log out with `Errorf`, so we can investigate when it does happen. https://github.com/juju/pubsub/pull/13/commits/f89121c392d9ebdd95131a0175cdba2329382c43

jameinel replied:
If you can do a quick sanity test once this lands, to make sure there isn't some unexpected condition where this is happening a lot. (do a bootstrap, deploy a couple things, make sure you don't see the new error message in the logs)

jameinel commented on metrics.go:
```diff
@@ -0,0 +1,44 @@
+	// Enqueued metric increments the number of messages a subscriber has
+	// currently. This can be used to see if a subscriber has a backlog of
+	// messages pilling up.
+	Enqueued(index int)
```
Instead of just index we could do something like:
https://stackoverflow.com/questions/7052693/how-to-get-the-name-of-a-function-in-go
and try to pass a string (that might fall back to str(index)). It would probably be slightly nicer for having meaningful metrics.

SimonRichardson replied:
Let me experiment with that one.

SimonRichardson replied:
https://github.com/juju/pubsub/pull/13/commits/80afb6d737b2c78681c61d4a751750ed12fd3aee

jameinel reviewed (commented):
I think this is certainly better than what we've had. I'd like to have sort of human meaningful identifier associated with the metrics so we can see which subscriber is being invalid.

As an aside to this PR, digging into the Raft Forwarder, it is listening on PubSub to get a request, then apply it to Raft and then send it back on PubSub.

But I don't see why we couldn't just pass the same Raft object into the lease engines and have them directly do the apply and FSMResponse.

```
	var empty raftlease.ForwardResponse
	start := time.Now()
	future := w.config.Raft.Apply([]byte(command), applyTimeout)
	if err := future.Error(); err != nil {
		return empty, errors.Trace(err)
	}
	w.metrics.record(start, "apply")
	respValue := future.Response()
	response, ok := respValue.(raftlease.FSMResponse)
	if !ok {
		return empty, errors.Errorf("expected an FSMResponse, got %T: %#v", respValue, respValue)
	}
	response.Notify(w.config.Target)
```

SimonRichardson committed:
Metrics: Expose function name

Rather than exposing the index of the subscriber, expose the name of the
callback. We can do this using the runtime.FuncForPC. Although not
ideal, we get the function name at subscribe time and not at publish
time, so shouldn't impact performance during publishing.

The getFunctionName attempts to return a stable function name depending
on the type of callback passed in. Functions, methods and high order
functions should be covered in tests.

Additionally unsubscribe just uses an anonymous function to simplify the
code, as we no longer require to create unsubscriber handlers for every
subscribe.

SimonRichardson committed:
Logging: Log out when we get a nil message.

SimonRichardson commented on simplehub_test.go:
```diff
@@ -255,8 +255,8 @@ func (*SimpleHubSuite) TestMetricsDequeue(c *gc.C) {
-	c.Assert(queue, gc.DeepEquals, map[int]int{
-		0: 1,
+	c.Assert(queue, gc.DeepEquals, map[string]int{
+		"TestMetricsDequeue.func1": 1,
```
Example of function name :point_up:

jameinel commented on subscriber.go:
```diff
@@ -132,10 +140,36 @@ func (s *subscriber) notify(call *handlerCallback) {
+	default:
+		name = strings.Join(parts[len(parts)-2:], ".")
+	}
+	// Ensure we remove potential suffixes from the name of the function.
```
It might help to have examples here of why we are doing this filtering.
